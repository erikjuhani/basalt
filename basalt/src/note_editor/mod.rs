pub mod ast;
mod cursor;
pub mod editor;
pub mod parser;
mod render;
mod rich_text;
mod state;
mod text_buffer;
mod virtual_document;

use ratatui::{
    crossterm::event::{KeyCode, KeyEvent, KeyModifiers},
    layout::Size,
};
pub use state::EditorState;
pub use text_buffer::TextBuffer;

use crate::{
    app::{calc_scroll_amount, ActivePane, Message as AppMessage, ScrollAmount},
    explorer,
    note_editor::editor::{EditMode, State, View},
    outline,
};

#[derive(Clone, Debug, PartialEq)]
pub enum Message {
    Save,
    SwitchPaneNext,
    SwitchPanePrevious,
    ToggleExplorer,
    ToggleOutline,
    ToggleView,
    EditView,
    ReadView,
    Exit,
    KeyEvent(KeyEvent),
    CursorUp,
    CursorLeft,
    CursorRight,
    CursorWordForward,
    CursorWordBackward,
    CursorDown,
    ScrollUp(ScrollAmount),
    ScrollDown(ScrollAmount),
    SetRow(usize),
    Delete,
}

pub fn update<'a>(
    message: &Message,
    screen_size: Size,
    state: &mut State,
) -> Option<AppMessage<'a>> {
    match message {
        Message::CursorLeft => state.cursor_left(),
        Message::CursorRight => state.cursor_right(),
        // TODO: Commented until editor functionality works
        // Message::CursorWordForward => state.cursor_word_forward(),
        // Message::CursorWordBackward => state.cursor_word_backward(),
        // Message::Delete => state.delete_char(),
        // Message::SetRow(row) => state.set_row(*row),
        Message::CursorUp => {
            state.cursor_up(1);
            return Some(AppMessage::Outline(outline::Message::SelectAt(
                state.current_row(),
            )));
        }
        Message::CursorDown => {
            state.cursor_down(1);
            return Some(AppMessage::Outline(outline::Message::SelectAt(
                state.current_row(),
            )));
        }
        Message::ScrollUp(scroll_amount) => {
            state.cursor_up(calc_scroll_amount(scroll_amount, screen_size.height.into()));
            return Some(AppMessage::Outline(outline::Message::SelectAt(
                state.current_row(),
            )));
        }
        Message::ScrollDown(scroll_amount) => {
            state.cursor_down(calc_scroll_amount(scroll_amount, screen_size.height.into()));
            return Some(AppMessage::Outline(outline::Message::SelectAt(
                state.current_row(),
            )));
        }
        _ => {}
    };

    match state.view {
        View::Edit(..) => match message {
            Message::ToggleView => state.set_view(View::Edit(EditMode::Source)),
            // TODO: Commented until editor functionality works
            //         Message::ScrollUp(_) => state.cursor_up(),
            //         Message::ScrollDown(_) => state.cursor_down(),
            //         Message::KeyEvent(key) => {
            //             state.edit((*key).into());
            //             return Some(AppMessage::UpdateSelectedNoteContent((
            //                 state.content().to_string(),
            //                 None,
            //             )));
            //         }
            Message::Exit => {
                // state.exit_insert();
                state.set_view(View::Read);
                // return Some(AppMessage::UpdateSelectedNoteContent((
                //     state.content().to_string(),
                //     Some(state.nodes().to_vec()),
                // )));
            }
            _ => {}
        },
        View::Read => match message {
            Message::ToggleView => state.set_view(View::Edit(EditMode::Source)),
            Message::EditView => state.set_view(View::Edit(EditMode::Source)),
            Message::ReadView => state.set_view(View::Read),
            Message::ToggleExplorer => {
                return Some(AppMessage::Explorer(explorer::Message::Toggle));
            }
            Message::ToggleOutline => {
                return Some(AppMessage::Outline(outline::Message::Toggle));
            }
            Message::SwitchPaneNext => {
                state.set_active(false);
                return Some(AppMessage::SetActivePane(ActivePane::Outline));
            }
            Message::SwitchPanePrevious => {
                state.set_active(false);
                return Some(AppMessage::SetActivePane(ActivePane::Explorer));
            }
            // TODO: Commented until editor functionality works
            //         Message::Save => {
            //             state.save();
            //             return Some(AppMessage::UpdateSelectedNoteContent((
            //                 state.content().to_string(),
            //                 None,
            //             )));
            //         }
            _ => {}
        },
    }

    None
}

pub fn handle_editing_event(key: &KeyEvent) -> Option<Message> {
    match key.code {
        KeyCode::Up => Some(Message::CursorUp),
        KeyCode::Down => Some(Message::CursorDown),
        KeyCode::Esc => Some(Message::Exit),
        KeyCode::Backspace => Some(Message::Delete),
        KeyCode::Char('e') if key.modifiers.contains(KeyModifiers::CONTROL) => {
            Some(Message::ToggleView)
        }
        _ => Some(Message::KeyEvent(*key)),
    }
}
