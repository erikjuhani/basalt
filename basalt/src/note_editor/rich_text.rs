use std::{fmt, slice::Iter, vec::IntoIter};

use pulldown_cmark::CowStr;

/// A style that can be applied to [`TextSegment`] (code, emphasis, strikethrough, strong).
#[derive(Clone, Debug, PartialEq)]
pub enum Style {
    Code,
    Emphasis,
    Strong,
    Strikethrough,
}

impl fmt::Display for Style {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Style::Code => write!(f, "Code"),
            Style::Emphasis => write!(f, "Emphasis"),
            Style::Strong => write!(f, "Strong"),
            Style::Strikethrough => write!(f, "Strikethrough"),
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct TextSegment {
    pub content: String,
    pub style: Option<Style>,
}

impl TextSegment {
    pub fn new(content: &str, style: Option<Style>) -> Self {
        Self {
            content: content.to_string(),
            style,
        }
    }

    pub fn add_style(&mut self, style: &Style) {
        self.style = Some(style.clone());
    }

    pub fn plain(content: &str) -> Self {
        Self {
            content: content.to_string(),
            style: None,
        }
    }

    pub fn styled(content: &str, style: Style) -> Self {
        Self {
            content: content.to_string(),
            style: Some(style),
        }
    }
}

impl fmt::Display for TextSegment {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let content = &self.content;
        write!(f, "{content}")
    }
}

impl From<CowStr<'_>> for TextSegment {
    fn from(content: CowStr<'_>) -> Self {
        content.to_string().into()
    }
}

impl From<String> for TextSegment {
    fn from(content: String) -> Self {
        content.as_str().into()
    }
}

impl From<&str> for TextSegment {
    fn from(content: &str) -> Self {
        // Replace tab character with spaces.
        // Tab character can break the rendered TUI.
        // TODO: Control tab length with config value.
        Self::plain(&content.replace("\t", "  "))
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct RichText {
    segments: Vec<TextSegment>,
}

impl IntoIterator for RichText {
    type Item = TextSegment;
    type IntoIter = IntoIter<Self::Item>;
    fn into_iter(self) -> Self::IntoIter {
        self.segments.into_iter()
    }
}

impl fmt::Display for RichText {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{}",
            self.segments
                .iter()
                .map(|segment| segment.to_string())
                .collect::<Vec<_>>()
                .join("")
        )
    }
}

impl RichText {
    pub fn empty() -> Self {
        Self::default()
    }

    pub fn iter(&self) -> Iter<'_, TextSegment> {
        self.segments.iter()
    }

    pub fn is_empty(&self) -> bool {
        self.segments.is_empty()
    }

    pub fn push_text_segment(&mut self, text_segment: TextSegment) {
        self.segments.push(text_segment);
    }

    pub fn segments(&self) -> &[TextSegment] {
        &self.segments
    }
}

impl From<Vec<TextSegment>> for RichText {
    fn from(segments: Vec<TextSegment>) -> Self {
        Self { segments }
    }
}

impl<const N: usize> From<[TextSegment; N]> for RichText {
    fn from(segments: [TextSegment; N]) -> Self {
        Self {
            segments: segments.to_vec(),
        }
    }
}
