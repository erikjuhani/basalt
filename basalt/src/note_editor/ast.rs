use crate::note_editor::rich_text::RichText;

#[derive(Clone, Copy, Debug, PartialEq, PartialOrd)]
pub enum HeadingLevel {
    H1 = 1,
    H2,
    H3,
    H4,
    H5,
    H6,
}

impl From<pulldown_cmark::HeadingLevel> for HeadingLevel {
    fn from(value: pulldown_cmark::HeadingLevel) -> Self {
        match value {
            pulldown_cmark::HeadingLevel::H1 => HeadingLevel::H1,
            pulldown_cmark::HeadingLevel::H2 => HeadingLevel::H2,
            pulldown_cmark::HeadingLevel::H3 => HeadingLevel::H3,
            pulldown_cmark::HeadingLevel::H4 => HeadingLevel::H4,
            pulldown_cmark::HeadingLevel::H5 => HeadingLevel::H5,
            pulldown_cmark::HeadingLevel::H6 => HeadingLevel::H6,
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum BlockQuoteKind {
    Note,
    Tip,
    Important,
    Warning,
    Caution,
}

impl From<pulldown_cmark::BlockQuoteKind> for BlockQuoteKind {
    fn from(value: pulldown_cmark::BlockQuoteKind) -> Self {
        match value {
            pulldown_cmark::BlockQuoteKind::Tip => BlockQuoteKind::Tip,
            pulldown_cmark::BlockQuoteKind::Note => BlockQuoteKind::Note,
            pulldown_cmark::BlockQuoteKind::Warning => BlockQuoteKind::Warning,
            pulldown_cmark::BlockQuoteKind::Caution => BlockQuoteKind::Caution,
            pulldown_cmark::BlockQuoteKind::Important => BlockQuoteKind::Important,
        }
    }
}

/// Denotes whether a list is ordered or unordered.
#[derive(Clone, Debug, PartialEq)]
pub enum ListKind {
    /// An ordered list item (e.g., `1. item`), storing the numeric index.
    Ordered(u64),
    /// An unordered list item (e.g., `- item`).
    Unordered,
}

/// Represents the variant of a list or task item (checked, unchecked, etc.).
#[derive(Clone, Debug, PartialEq)]
pub enum TaskKind {
    /// A checkbox item that is marked as done using `- [x]`.
    Checked,
    /// A checkbox item that is unchecked using `- [ ]`.
    Unchecked,
    /// A checkbox item that is checked, but not explicitly recognized as
    /// `Checked` (e.g., `- [?]`).
    LooselyChecked,
}

pub type SourceRange<Idx> = std::ops::Range<Idx>;

/// The Markdown AST node enumeration.
#[derive(Clone, Debug, PartialEq)]
pub enum Node {
    Heading {
        level: HeadingLevel,
        text: RichText,
        source_range: SourceRange<usize>,
    },
    Paragraph {
        text: RichText,
        source_range: SourceRange<usize>,
    },
    CodeBlock {
        lang: Option<String>,
        text: RichText,
        source_range: SourceRange<usize>,
    },
    BlockQuote {
        kind: Option<BlockQuoteKind>,
        nodes: Vec<Node>,
        source_range: SourceRange<usize>,
    },
    List {
        kind: ListKind,
        nodes: Vec<Node>,
        source_range: SourceRange<usize>,
    },
    Item {
        nodes: Vec<Node>,
        source_range: SourceRange<usize>,
    },
    Task {
        kind: TaskKind,
        nodes: Vec<Node>,
        source_range: SourceRange<usize>,
    },
}

impl Node {
    pub fn source_range(&self) -> &SourceRange<usize> {
        match self {
            Self::Heading { source_range, .. }
            | Self::CodeBlock { source_range, .. }
            | Self::Paragraph { source_range, .. }
            | Self::List { source_range, .. }
            | Self::BlockQuote { source_range, .. }
            | Self::Item { source_range, .. }
            | Self::Task { source_range, .. } => source_range,
        }
    }

    pub fn set_source_range(&mut self, new_range: SourceRange<usize>) {
        match self {
            Self::Heading { source_range, .. }
            | Self::CodeBlock { source_range, .. }
            | Self::Paragraph { source_range, .. }
            | Self::List { source_range, .. }
            | Self::BlockQuote { source_range, .. }
            | Self::Item { source_range, .. }
            | Self::Task { source_range, .. } => *source_range = new_range,
        }
    }
}

pub fn nodes_to_sexp(nodes: &[Node]) -> String {
    nodes
        .iter()
        .map(node_to_sexp)
        .collect::<Vec<_>>()
        .join("\n")
}

pub fn node_to_sexp(node: &Node) -> String {
    match node {
        Node::Heading {
            level,
            text,
            source_range,
        } => {
            format!(
                "(heading {:?} @{:?}\n  {})",
                level,
                source_range,
                rich_text_to_sexp(text)
            )
        }
        Node::Paragraph { text, source_range } => {
            format!(
                "(paragraph @{:?}\n  {})",
                source_range,
                rich_text_to_sexp(text)
            )
        }
        Node::BlockQuote {
            kind,
            nodes,
            source_range,
        } => {
            format!(
                "(blockquote {:?} @{:?}\n  {})",
                kind,
                source_range,
                nodes_to_sexp(nodes)
            )
        }
        Node::CodeBlock {
            lang,
            text,
            source_range,
        } => {
            format!(
                "(codeblock {} @{:?}\n  {})",
                lang.clone().unwrap_or_default(),
                source_range,
                rich_text_to_sexp(text)
            )
        }
        Node::List {
            kind,
            nodes,
            source_range,
        } => {
            format!(
                "(list {:?} @{:?}\n  {})",
                kind,
                source_range,
                nodes_to_sexp(nodes)
            )
        }
        Node::Item {
            nodes,
            source_range,
        } => {
            format!("(item @{:?}\n  {})", source_range, nodes_to_sexp(nodes))
        }
        Node::Task {
            kind,
            nodes,
            source_range,
        } => {
            format!(
                "(item {:?} @{:?}\n  {})",
                kind,
                source_range,
                nodes_to_sexp(nodes)
            )
        }
    }
}

pub fn rich_text_to_sexp(rich_text: &RichText) -> String {
    rich_text
        .segments()
        .iter()
        .map(|segment| match &segment.style {
            Some(style) => format!("({} \"{}\")", style, segment),
            None => format!("\"{}\"", segment),
        })
        .collect::<Vec<_>>()
        .join("\n  ")
}

// impl Node {
//     pub fn push_text_segment(&mut self, text_segment: TextSegment) {
//         match self {
//             Node::Heading { text, .. } => text.push_text_segment(text_segment),
//             Node::BlockQuote { nodes, .. } => {
//                 if let Some(last_node) = nodes.last_mut() {
//                     last_node.push_text_segment(text_segment);
//                 }
//             }
//             _ => {}
//         }
//     }
// }
