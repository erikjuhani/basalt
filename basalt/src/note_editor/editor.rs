use std::{fmt, marker::PhantomData, ops::Deref};

use ratatui::{
    buffer::Buffer,
    layout::{Offset, Rect, Size},
    style::{Color, Stylize},
    text::Line,
    widgets::{
        Block, BorderType, Padding, Paragraph, Scrollbar, ScrollbarOrientation, ScrollbarState,
        StatefulWidget, Widget,
    },
};

use crate::note_editor::{
    ast,
    cursor::{Cursor, CursorWidget},
    parser,
    virtual_document::VirtualDocument,
};

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub enum EditMode {
    #[default]
    /// Shows the markdown exactly as written
    Source,
    // TODO:
    // /// Hides most of the markdown syntax
    // LivePreview
}

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub enum View {
    #[default]
    Read,
    Edit(EditMode),
}

impl fmt::Display for View {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            View::Read => write!(f, "READ"),
            View::Edit(..) => write!(f, "EDIT"),
        }
    }
}

#[derive(Clone, Debug, Default)]
struct Viewport {
    area: Rect,
}

impl Deref for Viewport {
    type Target = Rect;

    fn deref(&self) -> &Self::Target {
        &self.area
    }
}

impl Viewport {
    pub fn resize(&mut self, size: Size) {
        self.area.width = size.width;
        self.area.height = size.height;
    }

    pub fn scroll_by(&mut self, offset: (i32, i32)) {
        self.area = self.offset(Offset {
            y: offset.0,
            x: offset.1,
        })
    }
}

// TODO: Some Editable block when user hits insert?
#[derive(Clone, Debug, Default)]
pub struct State<'a> {
    // TODO: Use Rope instead of String for O(log n) instead of O(n).
    content: String,
    pub view: View,
    pub cursor: Cursor,
    filename: String,
    pub ast_nodes: Vec<ast::Node>,
    virtual_document: VirtualDocument<'a>,
    active: bool,
    modified: bool,
    viewport: Viewport,
}

impl<'a> State<'a> {
    pub fn new(content: &str, filename: &str) -> Self {
        let ast_nodes = parser::from_str(content);
        let content = content.to_string();
        Self {
            content: content.clone(),
            view: View::Read,
            cursor: Cursor::default(),
            viewport: Viewport::default(),
            virtual_document: VirtualDocument::default(),
            filename: filename.to_string(),
            ast_nodes,
            active: false,
            modified: false,
        }
    }

    pub fn active(&self) -> bool {
        self.active
    }

    pub fn current_row(&self) -> usize {
        *self
            .virtual_document
            .line_to_block()
            .get(self.cursor.virtual_line())
            .unwrap_or(&0)
    }

    pub fn scroll_by(&mut self, offset: (i32, i32)) {
        self.viewport.scroll_by(offset);
    }

    pub fn set_view(&mut self, view: View) {
        self.view = view;

        self.virtual_document.layout(
            &self.filename,
            &self.content,
            &self.view,
            self.cursor.virtual_line(),
            &self.ast_nodes,
            self.viewport.width.into(),
        );

        match view {
            View::Read => self.cursor.enter_read_mode(&self.virtual_document),
            View::Edit(..) => self.cursor.enter_edit_mode(&self.virtual_document),
        }
    }

    pub fn resize_viewport(&mut self, size: Size) {
        // If width has changed we need to recalculate the wrapped lines.
        // Height change doesn't matter as it only affects what is visible.
        if self.viewport.width != size.width {
            self.virtual_document.layout(
                &self.filename,
                &self.content,
                &self.view,
                self.cursor.virtual_line(),
                &self.ast_nodes,
                size.width.into(),
            );
        }

        // TODO: if height has changed we need to move cursor if it goes out of bounds This means
        // we need to move it to last visible spot. We only need to care about the bottom()
        // boundary.
        self.viewport.resize(size);
    }

    pub fn set_active(&mut self, active: bool) {
        self.active = active;
    }

    pub fn cursor_left(&mut self) {}

    pub fn cursor_right(&mut self) {}

    // TODO: Applies to both cursor_up and cursor_down
    // The cursor should always be fixed to the viewport. This would enable easier implementation
    // for e.g. search feature when navigating between matches
    pub fn cursor_up(&mut self, amount: usize) {
        let prevline = self.cursor.virtual_line();
        self.cursor.cursor_up(amount, self.virtual_document.lines());

        let diff = self.cursor.virtual_line() as i32 - prevline as i32;

        if self.cursor.virtual_line() < self.viewport.top() as usize {
            self.viewport.scroll_by((diff, 0));
        }
    }

    pub fn cursor_down(&mut self, amount: usize) {
        // TODO: Implement scroll off so that the note scroll offset can be changed by moving
        // cursor downwards when we are at the bottom.
        let prevline = self.cursor.virtual_line();
        self.cursor
            .cursor_down(amount, self.virtual_document.lines());

        let diff = self.cursor.virtual_line() as i32 - prevline as i32;

        if self.cursor.virtual_line()
            >= self
                .viewport
                .bottom()
                .saturating_sub(self.virtual_document.meta().len() as u16) as usize
        {
            self.viewport.scroll_by((diff, 0));
        }
    }
}

#[derive(Default)]
pub struct NoteEditor<'a>(pub PhantomData<&'a ()>);

impl<'a> StatefulWidget for NoteEditor<'a> {
    type State = State<'a>;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let mode_color = match state.view {
            View::Edit(..) => Color::Green,
            View::Read => Color::Red,
        };

        let block = Block::bordered()
            .border_type(if state.active {
                BorderType::Thick
            } else {
                BorderType::Rounded
            })
            .title_bottom(
                [
                    format!(" {}", state.view).fg(mode_color).bold().italic(),
                    if state.modified {
                        "* ".bold().italic()
                    } else {
                        " ".into()
                    },
                ]
                .to_vec(),
            )
            .padding(Padding::horizontal(1));

        let inner_area = block.inner(area);

        // We only reliable know the size of the area for the editor once we arrive at this point.
        // Calling the resize_width will cause the visual_blocks to be populated in the state.
        // If width is not changed between frames the resize_width is a noop.
        state.resize_viewport(inner_area.as_size());

        let mut lines = state.virtual_document.meta().to_vec();
        lines.extend(state.virtual_document.lines().to_vec());

        let visible_lines = lines
            .iter()
            .skip(state.viewport.top() as usize)
            .take(state.viewport.bottom() as usize)
            // Cheaper to clone the subset of the lines
            .cloned()
            .map(|visual_line| visual_line.into())
            .collect::<Vec<Line>>();

        let rendered_lines_count = state.virtual_document.lines().len();
        let meta_lines_count = state.virtual_document.meta().len();

        Paragraph::new(visible_lines).block(block).render(area, buf);

        if !state.content.is_empty() {
            CursorWidget::default()
                .with_offset(Offset {
                    x: inner_area.x as i32,
                    y: inner_area.y as i32 + meta_lines_count as i32,
                })
                .render(state.viewport.area, buf, &mut state.cursor);
        }

        if !area.is_empty() && lines.len() as u16 > inner_area.bottom() {
            let mut scroll_state =
                ScrollbarState::new(rendered_lines_count).position(state.cursor.virtual_line());

            Scrollbar::new(ScrollbarOrientation::VerticalRight).render(
                area,
                buf,
                &mut scroll_state,
            );
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::note_editor::editor::EditMode;

    use super::*;
    use indoc::indoc;
    use insta::assert_snapshot;
    use ratatui::{
        backend::TestBackend,
        // crossterm::event::{KeyCode, KeyEvent, KeyModifiers},
        Terminal,
    };

    #[test]
    fn test_rendered_markdown_view() {
        let tests = [
            indoc! { r#"## Headings

            # This is a heading 1

            ## This is a heading 2

            ### This is a heading 3

            #### This is a heading 4

            ##### This is a heading 5

            ###### This is a heading 6
            "#},
            indoc! { r#"## Quotes

            You can quote text by adding a > symbols before the text.

            > Human beings face ever more complex and urgent problems, and their effectiveness in dealing with these problems is a matter that is critical to the stability and continued progress of society.
            >
            > - Doug Engelbart, 1961
            "#},
            indoc! { r#"## Callout Blocks

            > [!tip]
            >
            >You can turn your quote into a [callout](https://help.obsidian.md/Editing+and+formatting/Callouts) by adding `[!info]` as the first line in a quote.
            "#},
            indoc! { r#"## Deep Quotes

            You can have deeper levels of quotes by adding a > symbols before the text inside the block quote.

            > Regular thoughts
            >
            > > Deeper thoughts
            > >
            > > > Very deep thoughts
            > > >
            > > > - Someone on the internet 1996
            >
            > Back to regular thoughts
            "#},
            indoc! { r#"## Lists

            You can create an unordered list by adding a `-`, `*`, or `+` before the text.

            - First list item
            - Second list item
            - Third list item

            To create an ordered list, start each line with a number followed by a `.` symbol.

            1. First list item
            2. Second list item
            3. Third list item
            "#},
            indoc! { r#"## Indented Lists

            Lists can be indented

            - First list item
              - Second list item
                - Third list item

            "#},
            indoc! { r#"## Task lists

            To create a task list, start each list item with a hyphen and space followed by `[ ]`.

            - [x] This is a completed task.
            - [ ] This is an incomplete task.

            >You can use any character inside the brackets to mark it as complete.

            - [x] Oats
            - [?] Flour
            - [d] Apples
            "#},
            indoc! { r#"## Code blocks

            To format a block of code, surround the code with triple backticks.

            ```
            cd ~/Desktop
            ```

            You can also create a code block by indenting the text using `Tab` or 4 blank spaces.

                cd ~/Desktop
            "#},
            indoc! { r#"## Code blocks

            You can add syntax highlighting to a code block, by adding a language code after the first set of backticks.

            ```js
            function fancyAlert(arg) {
              if(arg) {
                $.facebox({div:'#foo'})
              }
            }
            ```
            "#},
        ];

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();

        tests.iter().for_each(|text| {
            _ = terminal.clear();
            let mut state = State::new(text, "Test");
            terminal
                .draw(|frame| {
                    NoteEditor::default().render(frame.area(), frame.buffer_mut(), &mut state)
                })
                .unwrap();
            assert_snapshot!(terminal.backend());
        });
    }

    #[test]
    fn test_rendered_editor_states() {
        let content = indoc! { r#"## Deep Quotes

            You can have deeper levels of quotes by adding a > symbols before the text inside the block quote.

            > Regular thoughts
            >
            > > Deeper thoughts
            > >
            > > > Very deep thoughts
            > > >
            > > > - Someone on the internet 1996
            >
            > Back to regular thoughts
            "#};

        let tests = [
            ("empty_default_state", State::default()),
            ("with_content", State::new(content, "Test")),
            ("read_mode_with_content", {
                let mut state = State::new(content, "Test");
                state.set_view(View::Read);
                state
            }),
            ("edit_mode_with_content", {
                let mut state = State::new(content, "Test");
                state.set_view(View::Edit(EditMode::Source));
                state
            }),
            // ("edit_mode_with_content_and_simple_change", {
            //     let mut state = State::new(content, "Test");
            //     state.set_view(View::Edit(EditMode::Source));
            //     state.edit(KeyEvent::new(KeyCode::Char('#'), KeyModifiers::empty()).into());
            //     state.exit_insert();
            //     state.set_view(View::Read);
            //     state
            // }),
            // ("edit_mode_with_arbitrary_cursor_move", {
            //     let mut state = State::new(content, "Test");
            //     state.set_content(content);
            //     state.cursor_move_col(7);
            //     state.set_view(View::Edit(EditMode::Source));
            //     state.edit(KeyEvent::new(KeyCode::Char(' '), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('B'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('a'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('s'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('a'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('l'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('t'), KeyModifiers::empty()).into());
            //     state.exit_insert();
            //     state.set_view(View::Read);
            //     state
            // }),
            // ("edit_mode_with_content_with_complete_word_input_change", {
            //     let mut state = State::new(content, "Test");
            //     state.set_content(content);
            //     state.cursor_down();
            //     state.set_view(View::Edit(EditMode::Source));
            //     state.edit(KeyEvent::new(KeyCode::Enter, KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('B'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('a'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('s'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('a'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('l'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Char('t'), KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Enter, KeyModifiers::empty()).into());
            //     state.edit(KeyEvent::new(KeyCode::Enter, KeyModifiers::empty()).into());
            //     state.exit_insert();
            //     state.set_view(View::Read);
            //     state
            // }),
        ];

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();

        tests.into_iter().for_each(|(name, mut state)| {
            _ = terminal.clear();
            terminal
                .draw(|frame| {
                    NoteEditor::default().render(frame.area(), frame.buffer_mut(), &mut state)
                })
                .unwrap();
            assert_snapshot!(name, terminal.backend());
        });
    }

    #[test]
    fn test_basic_formatting() {
        let tests = [
            (
                "paragraphs",
                indoc! { r#"## Paragraphs
                To create paragraphs in Markdown, use a **blank line** to separate blocks of text. Each block of text separated by a blank line is treated as a distinct paragraph.

                This is a paragraph.

                This is another paragraph.

                A blank line between lines of text creates separate paragraphs. This is the default behavior in Markdown.
                "#},
            ),
            (
                "headings",
                indoc! { r#"## Headings
                To create a heading, add up to six `#` symbols before your heading text. The number of `#` symbols determines the size of the heading.

                # This is a heading 1
                ## This is a heading 2
                ### This is a heading 3
                #### This is a heading 4
                ##### This is a heading 5
                ###### This is a heading 6
                "#},
            ),
            (
                "lists",
                indoc! { r#"## Lists
                You can create an unordered list by adding a `-`, `*`, or `+` before the text.

                - First list item
                - Second list item
                - Third list item

                To create an ordered list, start each line with a number followed by a `.` or `)` symbol.

                1. First list item
                2. Second list item
                3. Third list item

                1) First list item
                2) Second list item
                3) Third list item
                "#},
            ),
            (
                "lists_line_breaks",
                indoc! { r#"## Lists with line breaks
                You can use line breaks within an ordered list without altering the numbering.

                1. First list item

                2. Second list item
                3. Third list item

                4. Fourth list item
                5. Fifth list item
                6. Sixth list item
                "#},
            ),
            (
                "task_lists",
                indoc! { r#"## Task lists
                To create a task list, start each list item with a hyphen and space followed by `[ ]`.

                - [x] This is a completed task.
                - [ ] This is an incomplete task.

                You can toggle a task in Reading view by selecting the checkbox.

                > [!tip]
                > You can use any character inside the brackets to mark it as complete.
                >
                > - [x] Milk
                > - [?] Eggs
                > - [-] Eggs
                "#},
            ),
            (
                "nesting_lists",
                indoc! { r#"## Nesting lists
                You can nest any type of list—ordered, unordered, or task lists—under any other type of list.

                To create a nested list, indent one or more list items. You can mix list types within a nested structure:

                1. First list item
                   1. Ordered nested list item
                2. Second list item
                   - Unordered nested list item
                "#},
            ),
            (
                "nesting_task_lists",
                indoc! { r#"## Nesting task lists
                Similarly, you can create a nested task list by indenting one or more list items:

                - [ ] Task item 1
                 - [ ] Subtask 1
                - [ ] Task item 2
                 - [ ] Subtask 1
                "#},
            ),
            // TODO: Implement horizontal rule
            // (
            //     "horizontal_rule",
            //     indoc! { r#"## Horizontal rule
            //     You can use three or more stars `***`, hyphens `---`, or underscore `___` on its own line to add a horizontal bar. You can also separate symbols using spaces.
            //
            //     ***
            //     ****
            //     * * *
            //     ---
            //     ----
            //     - - -
            //     ___
            //     ____
            //     _ _ _
            //     "#},
            // ),
            (
                "code_blocks",
                indoc! { r#"## Code blocks
                To format code as a block, enclose it with three backticks or three tildes.

                ```md
                cd ~/Desktop
                ```

                You can also create a code block by indenting the text using `Tab` or 4 blank spaces.

                    cd ~/Desktop

                "#},
            ),
            (
                "code_syntax_highlighting_in_blocks",
                indoc! { r#"## Code syntax highlighting in blocks
                You can add syntax highlighting to a code block, by adding a language code after the first set of backticks.

                ```js
                function fancyAlert(arg) {
                  if(arg) {
                    $.facebox({div:'#foo'})
                  }
                }
                ```
                "#},
            ),
        ];

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();

        tests.into_iter().for_each(|(name, content)| {
            let mut state = State::new(content, name);
            _ = terminal.clear();
            terminal
                .draw(|frame| {
                    NoteEditor::default().render(frame.area(), frame.buffer_mut(), &mut state)
                })
                .unwrap();
            assert_snapshot!(name, terminal.backend());
        });
    }
}
