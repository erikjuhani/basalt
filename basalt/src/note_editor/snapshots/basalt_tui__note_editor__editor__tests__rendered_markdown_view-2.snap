---
source: basalt/src/note_editor/editor.rs
expression: terminal.backend()
---
"╭──────────────────────────────────────────────────────────────────────────────╮"
"│ Quotes                                                                       │"
"│ ──────────────────────────────────────────────────────────────────────────── │"
"│ You can quote text by adding a > symbols before the text.                    │"
"│                                                                              │"
"│ ┃ Human beings face ever more complex and urgent problems, and their         │"
"│ ┃ effectiveness in dealing with these problems is a matter that is critical  │"
"│ ┃ to the stability and continued progress of society.                        │"
"│ ┃                                                                            │"
"│ ┃ - Doug Engelbart, 1961                                                     │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"╰ READ ────────────────────────────────────────────────────────────────────────╯"
